operators/test/e2e/cmd/run/run.go

```diff
@@ -0,0 +1,561 @@
+		if err == io.EOF {
+			log.Info("Log stream ended")
+			return
+		}
+
+		streamStatus <- err
+	}
+}
+
+func (h *helper) kubectlApplyTemplate(templatePath string, templateParam interface{}, deleteOnExit bool) error {
+	outFilePath, err := h.renderTemplate(templatePath, templateParam)
+	if err != nil {
+		return err
+	}
+
+	if err := h.kubectl("apply", "-f", outFilePath); err != nil {
+		return err
+	}
+
+	if deleteOnExit {
+		h.addCleanupFunc(func() {
+			log.Info("Deleting resources", "file", outFilePath)
+			if err := h.kubectl("delete", "--all", "-f", outFilePath); err != nil {
+				log.Error(err, "Failed to delete resources", "file", outFilePath)
+			}
+		})
+	}
+
+	return nil
+}
+
+func (h *helper) kubectl(command string, args ...string) error {
+	return h.exec(h.kubectlWrapper.Command(command, args...))
+}
+
+func (h *helper) exec(cmd *command.Command) error {
+	ctx, cancelFunc := context.WithTimeout(context.Background(), h.commandTimeout)
+	defer cancelFunc()
+
+	log.V(1).Info("Executing command", "command", cmd)
+	out, err := cmd.Execute(ctx)
+	if err != nil {
+		// suppress the stacktrace when the command fails naturally
+		if _, ok := err.(*exec.ExitError); ok {
+			log.Info("Command returned error code", "command", cmd, "message", err.Error())
+		} else {
+			log.Error(err, "Command execution failed", "command", cmd)
+		}
+
+		fmt.Fprintln(os.Stderr, string(out))
+		return errors.Wrapf(err, "command failed: [%s]", cmd)
+	}
+
+	if log.V(1).Enabled() {
+		fmt.Println(string(out))
+	}
+
+	return nil
+}
+
+func (h *helper) renderTemplate(templatePath string, param interface{}) (string, error) {
+	tmpl, err := template.New(filepath.Base(templatePath)).Funcs(sprig.TxtFuncMap()).ParseFiles(templatePath)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to parse template at %s", templatePath)
+	}
+
+	outFilePath := filepath.Join(h.testOutDir, strings.Replace(templatePath, "/", "_", -1))
+	f, err := os.Create(outFilePath)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create file: %s", outFilePath)
+	}
+
+	defer f.Close()
+	if err := tmpl.Execute(f, param); err != nil {
+		return "", errors.Wrapf(err, "failed to render template to %s", outFilePath)
+	}
+
+	return outFilePath, nil
+}
+
+func (h *helper) addCleanupFunc(cf func()) {
+	h.cleanupFuncs = append(h.cleanupFuncs, cf)
+}
+
+func (h *helper) runCleanup() {
+	if h.skipCleanup {
+		log.Info("Skipping cleanup")
+		return
+	}
+
+	// run the cleanup functions in the reverse order they were added
+	for i := len(h.cleanupFuncs) - 1; i >= 0; i-- {
+		cf := h.cleanupFuncs[i]
+		cf()
+	}
+}
+
+func (h *helper) dumpEventLog() {
+	f, err := os.Open(h.eventLog)
+	if err != nil {
+		log.Error(err, "Failed to open event log", "path", h.eventLog)
+		return
+	}
+
+	log.Info("Cluster event log")
+	var buffer [1024]byte
+	if _, err := io.CopyBuffer(os.Stdout, f, buffer[:]); err != nil {
+		log.Error(err, "Failed to output event log")
+	}
```

Comment:
Something we do since https://github.com/elastic/cloud-on-k8s/pull/1395:
if there are non-running pods at the end of a (failed) test, get events for those pods, so we know why they are not running (eg. missing CPU).
I'm wondering how easy this will be to figure out from the dump of all events?
Maybe we should keep the events dump, but also display pods statuses so we can link failing pods to their events?

Reply:
Actually may be cleaner to list pods in the test code that fails. Eg. when we wait for 3 pods to be ready in a test, log the current pods statuses before failing the test.

Reply:
Yes, it would be nice if the test code could output context-specific failure details (in `test.Eventually` for example) but that's big change that I didn't want to take on in this PR. 

Only the failed events are logged right now and during my tests there weren't too many of them so it was fairly easy to figure out what happened. The output looks like the following:

```
{"message":"found no controllers for pod \"test-cross-ns-assoc-es-jfhfqpgkvk\"","kind":"PodDisruptionBudget","name":"test-cross-ns-assoc-es-default","namespace":"e2e-1kdkq-mercury","raw_event":{"metadata":{"name":"test-cross-ns-assoc-es-default.15b5d40a74af29b5","namespace":"e2e-1kdkq-mercury","selfLink":"/api/v1/namespaces/e2e-1kdkq-mercury/events/test-cross-ns-assoc-es-default.15b5d40a74af29b5","uid":"6bb87636-b1dc-11e9-897e-42010a840060","resourceVersion":"365","creationTimestamp":"2019-07-29T08:39:46Z"},"involvedObject":{"kind":"PodDisruptionBudget","namespace":"e2e-1kdkq-mercury","name":"test-cross-ns-assoc-es-default","uid":"6b7eb01b-b1dc-11e9-ac8b-42010a840012","apiVersion":"policy/v1beta1","resourceVersion":"4816"},"reason":"NoControllers","message":"found no controllers for pod \"test-cross-ns-assoc-es-jfhfqpgkvk\"","source":{"component":"controllermanager"},"firstTimestamp":"2019-07-29T08:39:46Z","lastTimestamp":"2019-07-29T08:39:46Z","count":1,"type":"Warning","eventTime":null,"reportingComponent":"","reportingInstance":""}}
```